src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -106,38 +112,60 @@ void UccDiscoveryPlugin::_validate_ucc_candidates(const UccCandidates& ucc_candi
 
     const auto& soft_key_constraints = table->soft_key_constraints();
 
-    // Skip already discovered UCCs.
-    if (std::any_of(soft_key_constraints.cbegin(), soft_key_constraints.cend(),
-                    [&column_id](const auto& key_constraint) {
-                      const auto& columns = key_constraint.columns();
-                      return columns.size() == 1 && *columns.cbegin() == column_id;
-                    })) {
-      message << " [skipped (already known) in " << candidate_timer.lap_formatted() << "]";
+    // Find UCC matching the candidate, if it already exists on the table.
+    const auto& existing_ucc = std::find_if(soft_key_constraints.cbegin(), soft_key_constraints.cend(),
+                                            [&column_id](const auto& key_constraint) {
+                                              const auto& columns = key_constraint.columns();
+
+                                              return columns.size() == 1 && *columns.cbegin() == column_id;
+                                            });
+
+    // Check if MVCC data tells us that the existing UCC is guaranteed to be still valid.
+    // If it is, we can skip the expensive revalidation of the UCC.
+    if (existing_ucc != soft_key_constraints.cend() && table->constraint_guaranteed_to_be_valid(*existing_ucc)) {
+      message << " [skipped (already known and guaranteed to be still valid) in " << candidate_timer.lap_formatted()
+              << "]";
       Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+      existing_ucc->revalidated_on(transaction_context->snapshot_commit_id());
```

Comment:
Wait a sec. You can end up setting the validation snapshot ID for a schema-provided constraint here. (i) This makes the constraint not schema-given anymore (the check for this is equality w/ special commit ID). (ii) Your `DebugAssert` in `TableKeyConstraint` (ensure new commit > last validated commit) should throw if you do this.